File: triangle.cpp
//------------------------------------------------------------------------------
// triangle.cpp - содержит функции обработки треугольника
//------------------------------------------------------------------------------

#include "triangle.h"
#include "math.h"

//------------------------------------------------------------------------------
// Ввод параметров треугольника из потока
void In(triangle &t, ifstream &ifst) {
    ifst >> t.xa >> t.ya >> t.xb >> t.yb >> t.xc >> t.yc;
    t.bc_len = sqrt(pow(t.xc - t.xb, 2) + pow(t.yc - t.yb, 2));
    t.ac_len = sqrt(pow(t.xa - t.xc, 2) + pow(t.ya - t.yc, 2));
    t.ab_len = sqrt(pow(t.xa - t.xb, 2) + pow(t.ya - t.yb, 2));
}

// Случайный ввод параметров треугольника
void InRnd(triangle &t) {
    double ab_len, bc_len, ac_len;
    t.xa = Random();
    t.ya = Random();
    t.xb = Random();
    t.yb = Random();
    t.ab_len = sqrt(pow(t.xa - t.xb, 2) + pow(t.ya - t.yb, 2));
    do {
        t.yc = Random();
        t.xc = Random();
        t.bc_len = sqrt(pow(t.xc - t.xb, 2) + pow(t.yc - t.yb, 2));
        t.ac_len = sqrt(pow(t.xa - t.xc, 2) + pow(t.ya - t.yc, 2));
    } while((t.ab_len >= (t.bc_len + t.ac_len))
            && (t.bc_len >= (t.ab_len + t.ac_len))
            && (t.ac_len >= (t.ab_len + t.bc_len)));
}

//------------------------------------------------------------------------------
// Вывод параметров треугольника в поток
void Out(triangle &t, ofstream &ofst) {
    ofst << "It is Triangle with coordinates: A = ("
         << t.xa << ";" << t.ya << ")"
         << ", B = (" << t.xb << ";" << t.yb << ")"
         << ", C = (" << t.xc << ";" << t.yc << ")"
         << ". Perimeter = " << Perimeter(t) << "\n";
}

//------------------------------------------------------------------------------
// Вычисление периметра треугольника
double Perimeter(triangle &t) {
    return double(t.ab_len + t.ab_len + t.bc_len);
}

File: triangle.py
import math
import random

from shape import *


# ----------------------------------------------
class Triangle(Shape):
    def __init__(self):
        self.xa = 0
        self.ya = 0
        self.xb = 0
        self.yb = 0
        self.xc = 0
        self.yc = 0
        self.bc_len = math.sqrt(pow(self.xc - self.xb, 2) + pow(self.yc - self.yb, 2))
        self.ac_len = math.sqrt(pow(self.xa - self.xc, 2) + pow(self.ya - self.yc, 2))
        self.ab_len = math.sqrt(pow(self.xa - self.xb, 2) + pow(self.ya - self.yb, 2))

    def ReadStrArray(self, strArray, i):
        # должно быт как минимум 6 непрочитанных значения в массиве
        if i >= len(strArray) - 5:
            return 0
        self.xa = int(strArray[i])
        self.ya = int(strArray[i + 1])
        self.xb = int(strArray[i + 2])
        self.yb = int(strArray[i + 3])
        self.xc = int(strArray[i + 4])
        self.yc = int(strArray[i + 5])
        i += 6
        return i

    def Print(self):
        print("Triangle: a = (", self.xa, ";", self.ya, ") b = (", self.xb, ";", self.yb, ") c = (", self.xc, ";", self.yc, "), Perimeter = ",
              self.Perimeter())
        pass

    def Write(self, ostream):
        ostream.write("Triangle: a = ({};{})  b = ({};{})  c = ({};{}), Perimeter = {}"
                      .format(self.xa, self.ya, self.xb, self.yb, self.xc, self.yc, self.Perimeter()))
        pass

    def Perimeter(self):
        return float(self.ab_len + self.bc_len + self.ac_len)
        pass

    def FillRdn(self):
        self.xa = random.randint(1, 20)
        self.ya = random.randint(1, 20)
        self.xb = random.randint(1, 20)
        self.yb = random.randint(1, 20)
        self.ab_len = math.sqrt(pow(self.xa - self.xb, 2) + pow(self.ya - self.yb, 2))
        self.yc = random.randint(1, 20)
        self.xc = random.randint(1, 20)
        self.bc_len = math.sqrt(pow(self.xc - self.xb, 2) + pow(self.yc - self.yb, 2))
        self.ac_len = math.sqrt(pow(self.xa - self.xc, 2) + pow(self.ya - self.yc, 2))
        while (self.ab_len >= (self.bc_len + self.ac_len)) and (self.bc_len >= (self.ab_len + self.ac_len)) and (self.ac_len >= (self.ab_len + self.bc_len)):
            self.yc = random.randint(1, 20)
            self.xc = random.randint(1, 20)
        pass

